solarforecastarbiter/utils.py

```diff
@@ -0,0 +1,163 @@
+    """
+    nindex = pd.DatetimeIndex([], tz=index.tz)
+    for aggobs in aggregate_observations:
+        if aggobs['observation_id'] == obs_id:
+            if aggobs['observation_deleted_at'] is None:
+                locs = index.slice_locs(aggobs['effective_from'],
+                                        aggobs['effective_until'])
+                nindex = nindex.union(index[locs[0]:locs[1]])
+            elif (
+                    aggobs['effective_until'] is None or
+                    aggobs['effective_until'] >= index[0]
+            ):
+                raise ValueError(
+                    'Deleted Observation data cannot be retrieved'
+                    ' to include in Aggregate')
+            else:  # observation deleted and effective_until before index
+                return pd.Series(False, index=index)
+    return pd.Series(1, index=nindex).reindex(index).fillna(0).astype(bool)
+
+
+def _make_aggregate_index(data, interval_length, interval_label,
+                          timezone):
+    """
+    Compute the aggregate the index should have based on the min and
+    max timestamps in the data, the interval length, label, and timezone.
+    """
+    # first, find limits for a new index
+    start = pd.Timestamp('20380119T031407Z')
+    end = pd.Timestamp('19700101T000001Z')
+    for df in data.values():
+        start = min(start, min(df.index))
+        end = max(end, max(df.index))
+    # adjust start, end to nearest interval
+    # hard to understand what this interval should be for
+    # odd (e.g. 52min) intervals, so required that interval
+    # is a divisor of one day
+    if 86400 % pd.Timedelta(interval_length).total_seconds() != 0:
+        raise ValueError(
+            'interval_length must be a divisor of one day')
+    if interval_label == 'ending':
+        start = start.ceil(interval_length)
+        end = end.ceil(interval_length)
+    elif interval_label == 'beginning':
+        start = start.floor(interval_length)
+        end = end.floor(interval_length)
+    else:
+        raise ValueError(
+            'interval_label must be beginning or ending for aggregates')
+    # raise the error if unlocalized
+    start = start.tz_convert(timezone)
+    end = end.tz_convert(timezone)
+    return pd.date_range(
+        start, end, freq=interval_length, tz=timezone)
+
+
+def compute_aggregate(data, interval_length, interval_label,
+                      timezone, agg_func, aggregate_observations):
+    """
+    Computes an aggregate quantity according to agg_func of the data.
+    This function assumes the data has an interval_label of
+    interval_mean or instantaneous and that the data interva_length
+    is less than or equal to the aggregate interval_length.
+    NaNs in the output are the result of missing data from an
+    underyling observation of the aggregate.
+
+    Parameters
+    ----------
+    data : dict of pandas.DataFrames
+       With keys 'observation_id' corresponding to observation in
```

Comment:
Right now, this isn't really set up to handle the datamodel objects directly, since that would be somewhat unnecessary to do in the API where this function will be used. For datamodel objects, you would only really need the Aggregate object and the data anyway.